File: local-flags-certificates/sec_search/fast_check.py
#!/usr/bin/env python3
"""Fast χ'_s ≤ K check using greedy DSATUR pre-filter + SAT fallback.

Pipeline:
  1. Trivial: if L(G)² is complete, χ'_s = |E(G)|. Compare to K.
  2. ν_s lower bound: χ'_s ≥ ⌈|E|/ν_s⌉. If this > K, χ'_s > K (counterexample).
     (Cheap upper bound on ν_s via greedy maximal independent set in L²(G).)
  3. Clique lower bound: max found clique in L²(G); if > K, counterexample.
  4. DSATUR greedy upper bound on χ(L²(G)). If ≤ K, χ'_s ≤ K (pass).
  5. SAT fallback.

For 4-regular small-n graphs, ~99%+ are resolved by step 4 without
SAT. The few hard cases (typically high-girth, near-extremal) still
SAT-solve, but they're rare so the amortised cost is dramatically lower.
"""

import random
import sys
from itertools import combinations
from pathlib import Path

import networkx as nx

sys.path.insert(0, str(Path(__file__).resolve().parent))
from compute_chi_s import (strong_adjacency_pairs, is_k_strong_colourable,
                           _solve_k_colouring)


def build_L_squared(G: nx.Graph):
    """Return (edges_list, adj_set_per_edge) for L(G)²."""
    edges, pairs = strong_adjacency_pairs(G)
    n_e = len(edges)
    adj = [set() for _ in range(n_e)]
    for i, j in pairs:
        adj[i].add(j)
        adj[j].add(i)
    return edges, adj, pairs


def dsatur_colour(n: int, adj: list, max_colours: int = None) -> int:
    """DSATUR greedy chromatic upper bound for a graph on n vertices.

    adj: list of sets; adj[v] = neighbours of v.
    max_colours: early-abort cap (return None if would exceed).

    Returns the number of colours used by DSATUR greedy.
    """
    if n == 0:
        return 0
    deg = [len(a) for a in adj]
    colour = [-1] * n
    sat = [set() for _ in range(n)]
    sat_count = [0] * n
    coloured = 0
    while coloured < n:
        # Pick uncoloured vertex with max saturation (ties broken by degree).
        best = -1
        best_key = (-1, -1)
        for v in range(n):
            if colour[v] >= 0:
                continue
            key = (sat_count[v], deg[v])
            if key > best_key:
                best_key = key
                best = v
        # Lowest colour not in sat[best]
        c = 0
        while c in sat[best]:
            c += 1
        if max_colours is not None and c >= max_colours:
            return None  # exceeded cap, abort
        colour[best] = c
        for w in adj[best]:
            if colour[w] < 0 and c not in sat[w]:
                sat[w].add(c)
                sat_count[w] += 1
        coloured += 1
    return max(colour) + 1


def dsatur_colour_with_order(n: int, adj: list, order: list,
                              max_colours: int = None) -> int:
    """DSATUR-like greedy with a fixed vertex order (first-fit)."""
    if n == 0:
        return 0
    colour = [-1] * n
    for v in order:
        forbidden = {colour[u] for u in adj[v] if colour[u] >= 0}
        c = 0
        while c in forbidden:
            c += 1
        if max_colours is not None and c >= max_colours:
            return None
        colour[v] = c
    return max(colour) + 1


def greedy_chi_upper_bound(n: int, adj: list, max_colours: int,
                            attempts: int = 5, rng: random.Random = None) -> int:
    """Multi-shot greedy upper bound on χ(L²). Returns min over attempts."""
    if rng is None:
        rng = random.Random(0)
    best = None
    # First: DSATUR with degree-saturation ordering
    r = dsatur_colour(n, adj, max_colours=max_colours)
    if r is not None and (best is None or r < best):
        best = r
        if best <= max_colours:
            # Try a few more to see if we can do better, but we don't need
            # to — bound already satisfied.
            return best
    # Then: random orderings
    for _ in range(attempts):
        order = list(range(n))
        rng.shuffle(order)
        r = dsatur_colour_with_order(n, adj, order, max_colours=max_colours)
        if r is not None and (best is None or r < best):
            best = r
            if best <= max_colours:
                return best
    return best if best is not None else max_colours + 1


def greedy_max_indep_set(n: int, adj: list) -> int:
    """Greedy lower bound on α(L²(G)) = ν_s(G). Pick vertex of min deg,
    add to IS, remove it and its neighbours, repeat."""
    if n == 0:
        return 0
    alive = set(range(n))
    deg_alive = {v: len(adj[v] & alive) for v in alive}
    is_size = 0
    while alive:
        v = min(alive, key=lambda x: deg_alive[x])
        is_size += 1
        for u in list(adj[v]):
            if u in alive:
                alive.discard(u)
                for w in adj[u]:
                    if w in alive:
                        deg_alive[w] -= 1
        alive.discard(v)
    return is_size


def is_chi_s_at_most_fast(G: nx.Graph, k: int,
                          stats: dict = None) -> bool:
    """Fast check: χ'_s(G) ≤ k?

    Optionally records which decision path was used in stats dict
    (keys: 'shortcut_complete', 'nu_s_counterex', 'greedy_pass', 'sat_pass',
     'sat_fail').
    """
    edges, adj, pairs = build_L_squared(G)
    n_e = len(edges)
    if n_e == 0:
        return True

    # 1. L²(G) complete shortcut
    all_pairs = n_e * (n_e - 1) // 2
    if len(pairs) == all_pairs:
        result = n_e <= k
        if stats is not None:
            stats['shortcut_complete'] = stats.get('shortcut_complete', 0) + 1
        return result

    # 2. ν_s lower-bound counterexample check (cheap greedy)
    nu_s_lb = greedy_max_indep_set(n_e, adj)  # ≤ ν_s
    # χ'_s ≥ ⌈|E|/ν_s⌉. ν_s ≤ greedy result? NO — greedy gives a LOWER bound on ν_s.
    # So |E|/(greedy ν_s lb) ≥ |E|/ν_s, i.e., this is an UPPER bound on the LB,
    # not useful for proving counterexample.
    # Skip — ν_s greedy alone can't prove counterexample without exact ν_s.

    # 3. Greedy upper bound on χ(L²)
    ub = greedy_chi_upper_bound(n_e, adj, max_colours=k)
    if ub is not None and ub <= k:
        if stats is not None:
            stats['greedy_pass'] = stats.get('greedy_pass', 0) + 1
        return True

    # 4. SAT fallback
    sat = is_k_strong_colourable(n_e, pairs, k)
    if stats is not None:
        key = 'sat_pass' if sat else 'sat_fail'
        stats[key] = stats.get(key, 0) + 1
    return sat


if __name__ == "__main__":
    # Self-test: rerun a small panel and check answers match compute_chi_s.
    import time
    from compute_chi_s import (G_complete, G_complete_bipartite, G_cycle,
                                G_petersen, chi_s, G_5blowup_C5)

    tests = [
        ("K_4", G_complete(4), 6),
        ("K_5", G_complete(5), 10),
        ("C_5", G_cycle(5), 5),
        ("K_{3,3}", G_complete_bipartite(3, 3), 9),
        ("K_{4,4}", G_complete_bipartite(4, 4), 16),
        ("Petersen", G_petersen(), 5),
        ("5-blowup C_5", G_5blowup_C5(), 125),
    ]
    stats = {}
    for name, G, expected in tests:
        # Test ≤ expected (should be True) and ≤ expected-1 (depends)
        ok_at_exp = is_chi_s_at_most_fast(G, expected, stats)
        ok_below = is_chi_s_at_most_fast(G, expected - 1, stats)
        print(f"{name:14s} chi_s={expected}  k=exp: {ok_at_exp}  k=exp-1: {ok_below}")
    print(f"\nDecision-path stats: {stats}")

File: local-flags-certificates/sec_search/fast_check.c
/* fast_check.c — fast greedy χ'_s ≤ bound check via DSATUR.
 *
 * Reads graph6-encoded graphs from stdin (one per line), runs the
 * same pipeline as fast_check.py:
 *   1. L²(G)-complete shortcut → χ'_s = |E|.
 *   2. DSATUR greedy with degree-saturation + multi-shot random orderings.
 *      If colouring ≤ bound exists, PASS.
 *   3. Otherwise: report CANDIDATE g6=... (left to SAT downstream).
 *
 * Writes per-graph status to stdout in CSV:
 *   graph6,n,m,delta,girth_dummy,chi_s_le_bound,decision_path
 *
 * Aggregate stats to stderr.
 *
 * Compile: clang -O3 -march=native -o fast_check fast_check.c
 * Use:     geng -d4 -D4 -c 16 | ./fast_check 20
 *
 * Supports graphs with up to 64 edges (i.e. |V(L²)| ≤ 64), which
 * covers all 4-regular n ≤ 32 (|E|=64), 5-regular n ≤ 25 (|E|=62.5).
 * For larger graphs, fall back to Python (fast_check.py) or extend to
 * multi-word bitsets.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <time.h>

#define MAX_N 64       /* graph vertices */
#define MAX_M 512      /* edges / colours — 8-word bitset for Δ up to ~22 */
#define MAX_SHOTS 32   /* DSATUR random restarts (cap) */
#define EBS_WORDS 8    /* 8 * 64 = 512 bits */

/* Vertex bitsets (over n ≤ 64 vertices) — single word. */
typedef uint64_t bitset_t;

/* Edge / colour bitsets — 256-bit (4 × uint64_t). */
typedef struct {
    uint64_t w[EBS_WORDS];
} ebitset_t;

static inline ebitset_t ebs_zero(void) {
    ebitset_t r;
    for (int i = 0; i < EBS_WORDS; i++) r.w[i] = 0;
    return r;
}
static inline ebitset_t ebs_singleton(int bit) {
    ebitset_t r = ebs_zero();
    r.w[bit >> 6] = (uint64_t)1 << (bit & 63);
    return r;
}
static inline ebitset_t ebs_or(ebitset_t a, ebitset_t b) {
    for (int i = 0; i < EBS_WORDS; i++) a.w[i] |= b.w[i];
    return a;
}
static inline void ebs_or_inplace(ebitset_t *a, ebitset_t b) {
    for (int i = 0; i < EBS_WORDS; i++) a->w[i] |= b.w[i];
}
static inline ebitset_t ebs_not(ebitset_t a) {
    for (int i = 0; i < EBS_WORDS; i++) a.w[i] = ~a.w[i];
    return a;
}
static inline int ebs_is_zero(ebitset_t a) {
    uint64_t x = 0;
    for (int i = 0; i < EBS_WORDS; i++) x |= a.w[i];
    return x == 0;
}
static inline int ebs_test(ebitset_t a, int bit) {
    return (a.w[bit >> 6] >> (bit & 63)) & 1;
}
static inline int ebs_popcount(ebitset_t a) {
    int s = 0;
    for (int i = 0; i < EBS_WORDS; i++) s += __builtin_popcountll(a.w[i]);
    return s;
}
static inline int ebs_ctz(ebitset_t a) {
    for (int i = 0; i < EBS_WORDS; i++) {
        if (a.w[i]) return i * 64 + __builtin_ctzll(a.w[i]);
    }
    return EBS_WORDS * 64;
}
/* Clear the lowest set bit (equivalent to a &= a - 1). */
static inline ebitset_t ebs_clear_lowest(ebitset_t a) {
    for (int i = 0; i < EBS_WORDS; i++) {
        if (a.w[i]) { a.w[i] &= a.w[i] - 1; return a; }
    }
    return a;
}
/* Test whether two edges (vertex-side) intersect: any common bit. */
static inline int bs_intersects(bitset_t a, bitset_t b) { return (a & b) != 0; }

/* ---------------- graph6 parser ---------------- */

/* Parse graph6 line into adj[]: adj[v] is a bitset of v's neighbours
 * (assuming n ≤ MAX_N). Returns n on success, -1 on parse failure. */
static int parse_graph6(const char *line, bitset_t adj[MAX_N]) {
    if (!line || !*line) return -1;
    int n;
    const char *p = line;
    if ((unsigned char)*p < 63) return -1;
    /* n encoding */
    if (*p < 126) {
        n = *p - 63;
        p++;
    } else {
        /* multi-byte n encoding — not supported here (n>62) */
        return -1;
    }
    if (n < 0 || n > MAX_N) return -1;

    for (int v = 0; v < n; v++) adj[v] = 0;

    /* Edges encoded column-major: for j = 1..n-1, for i = 0..j-1,
     * one bit. Each char = 6 bits (high bit first). */
    int bit_idx = 0;
    int total_bits = n * (n - 1) / 2;
    int byte_bits = 0;
    int byte_val = 0;
    for (int j = 1; j < n; j++) {
        for (int i = 0; i < j; i++) {
            if (byte_bits == 0) {
                if (*p < 63) return -1;
                byte_val = *p - 63;
                p++;
                byte_bits = 6;
            }
            byte_bits--;
            int b = (byte_val >> byte_bits) & 1;
            if (b) {
                adj[i] |= ((bitset_t)1) << j;
                adj[j] |= ((bitset_t)1) << i;
            }
            bit_idx++;
        }
    }
    (void)bit_idx;
    (void)total_bits;
    return n;
}

/* ---------------- enumerate edges ---------------- */

/* Fill edges[].u and edges[].v (canonical u<v) from adj[]. Returns m. */
typedef struct { int u, v; } edge_t;

static int extract_edges(int n, const bitset_t adj[MAX_N], edge_t edges[MAX_M]) {
    int m = 0;
    for (int u = 0; u < n; u++) {
        bitset_t hi = adj[u] & ~((((bitset_t)1) << (u + 1)) - 1);
        while (hi) {
            int v = __builtin_ctzll(hi);
            hi &= hi - 1;
            if (m >= MAX_M) return -1;
            edges[m].u = u;
            edges[m].v = v;
            m++;
        }
    }
    return m;
}

/* ---------------- build L²(G) adjacency (bitset) ---------------- */

/* sadj[e] = bitmask of edges strong-adjacent to e (over up to MAX_M edges). */
static void build_L_squared(int n, const bitset_t adj[MAX_N],
                            int m, const edge_t edges[MAX_M],
                            ebitset_t sadj[MAX_M]) {
    bitset_t reach[MAX_M];
    for (int e = 0; e < m; e++) {
        int u = edges[e].u, v = edges[e].v;
        reach[e] = adj[u] | adj[v] | (((bitset_t)1) << u) | (((bitset_t)1) << v);
        sadj[e] = ebs_zero();
    }
    for (int e = 0; e < m; e++) {
        for (int f = e + 1; f < m; f++) {
            int x = edges[f].u, y = edges[f].v;
            bitset_t fmask = (((bitset_t)1) << x) | (((bitset_t)1) << y);
            if (reach[e] & fmask) {
                ebs_or_inplace(&sadj[e], ebs_singleton(f));
                ebs_or_inplace(&sadj[f], ebs_singleton(e));
            }
        }
    }
}

/* ---------------- DSATUR ---------------- */

/* DSATUR greedy with saturation-degree ordering. Returns the number
 * of colours used, or -1 if would exceed max_colours.
 * Operates on a graph of m vertices with adjacency sadj[i] (bitset). */
static int dsatur_colour(int m, const ebitset_t sadj[MAX_M], int max_colours) {
    int colour[MAX_M];
    ebitset_t sat[MAX_M];     /* forbidden COLOUR-bitset (up to 256 colours) */
    int sat_count[MAX_M];
    int deg[MAX_M];
    for (int v = 0; v < m; v++) {
        colour[v] = -1;
        sat[v] = ebs_zero();
        sat_count[v] = 0;
        deg[v] = ebs_popcount(sadj[v]);
    }
    int used = 0;
    for (int round = 0; round < m; round++) {
        int best = -1, best_sat = -1, best_deg = -1;
        for (int v = 0; v < m; v++) {
            if (colour[v] >= 0) continue;
            if (sat_count[v] > best_sat ||
                (sat_count[v] == best_sat && deg[v] > best_deg)) {
                best = v;
                best_sat = sat_count[v];
                best_deg = deg[v];
            }
        }
        ebitset_t inverted = ebs_not(sat[best]);
        if (ebs_is_zero(inverted)) return -1;  /* all 256 colours forbidden */
        int c = ebs_ctz(inverted);
        if (c >= max_colours) return -1;
        colour[best] = c;
        if (c + 1 > used) used = c + 1;
        ebitset_t nbrs = sadj[best];
        while (!ebs_is_zero(nbrs)) {
            int w = ebs_ctz(nbrs);
            nbrs = ebs_clear_lowest(nbrs);
            if (colour[w] < 0) {
                if (!ebs_test(sat[w], c)) {
                    ebs_or_inplace(&sat[w], ebs_singleton(c));
                    sat_count[w]++;
                }
            }
        }
    }
    return used;
}

/* First-fit greedy with explicit order. Returns colours used or -1. */
static int firstfit_colour(int m, const ebitset_t sadj[MAX_M],
                            const int order[MAX_M], int max_colours) {
    int colour[MAX_M];
    for (int i = 0; i < m; i++) colour[i] = -1;
    int used = 0;
    for (int idx = 0; idx < m; idx++) {
        int v = order[idx];
        ebitset_t forbidden = ebs_zero();
        ebitset_t nbrs = sadj[v];
        while (!ebs_is_zero(nbrs)) {
            int w = ebs_ctz(nbrs);
            nbrs = ebs_clear_lowest(nbrs);
            if (colour[w] >= 0) ebs_or_inplace(&forbidden, ebs_singleton(colour[w]));
        }
        ebitset_t inv = ebs_not(forbidden);
        if (ebs_is_zero(inv)) return -1;
        int c = ebs_ctz(inv);
        if (c >= max_colours) return -1;
        colour[v] = c;
        if (c + 1 > used) used = c + 1;
    }
    return used;
}

/* xorshift64 fast PRNG */
static uint64_t rng_state = 0x9E3779B97F4A7C15ULL;
static inline uint64_t rng_next() {
    uint64_t x = rng_state;
    x ^= x << 13; x ^= x >> 7; x ^= x << 17;
    rng_state = x;
    return x;
}

/* Multi-shot greedy: DSATUR first, then up to (shots-1) random first-fit.
 * Returns min colours used across attempts that succeed at ≤ max_colours,
 * or -1 if none succeed. */
static int multi_shot_greedy(int m, const ebitset_t sadj[MAX_M],
                              int max_colours, int shots) {
    int best = dsatur_colour(m, sadj, max_colours);
    if (best > 0 && best <= max_colours) return best;
    int order[MAX_M];
    for (int s = 1; s < shots; s++) {
        for (int i = 0; i < m; i++) order[i] = i;
        /* Fisher-Yates */
        for (int i = m - 1; i > 0; i--) {
            int j = (int)(rng_next() % (uint64_t)(i + 1));
            int tmp = order[i]; order[i] = order[j]; order[j] = tmp;
        }
        int r = firstfit_colour(m, sadj, order, max_colours);
        if (r > 0 && (best < 0 || r < best)) {
            best = r;
            if (best <= max_colours) return best;
        }
    }
    return best;
}

/* ---------------- main loop ---------------- */

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr,
            "Usage: %s <bound> [--csv] [--every N] [--shots N]\n"
            "Reads graph6 from stdin. Writes CSV/log to stdout/stderr.\n",
            argv[0]);
        return 1;
    }
    int bound = atoi(argv[1]);
    int csv = 0, shots = 20;
    long every = 100000;
    for (int i = 2; i < argc; i++) {
        if (strcmp(argv[i], "--csv") == 0) csv = 1;
        else if (strcmp(argv[i], "--every") == 0 && i + 1 < argc)
            every = atol(argv[++i]);
        else if (strcmp(argv[i], "--shots") == 0 && i + 1 < argc)
            shots = atoi(argv[++i]);
    }
    if (csv) {
        printf("graph6,n,m,bound,passed_path\n");
        fflush(stdout);
    }

    char line[8192];
    long count = 0;
    long candidates = 0;
    long path_shortcut = 0, path_greedy = 0, path_unresolved = 0;
    bitset_t adj[MAX_N];
    edge_t edges[MAX_M];
    ebitset_t sadj[MAX_M];

    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    while (fgets(line, sizeof(line), stdin)) {
        /* strip trailing newline */
        size_t L = strlen(line);
        while (L > 0 && (line[L-1] == '\n' || line[L-1] == '\r'))
            line[--L] = '\0';
        if (L == 0) continue;

        int n = parse_graph6(line, adj);
        if (n < 0) {
            fprintf(stderr, "# parse error: %s\n", line);
            continue;
        }
        int m = extract_edges(n, adj, edges);
        if (m < 0) {
            fprintf(stderr, "# too many edges (%d > %d): %s\n",
                    m, MAX_M, line);
            continue;
        }
        const char *path;
        int passed;
        if (m == 0) {
            passed = 1; path = "empty"; path_shortcut++;
        } else {
            build_L_squared(n, adj, m, edges, sadj);
            /* count strong-pairs */
            long s_pairs = 0;
            for (int e = 0; e < m; e++)
                s_pairs += ebs_popcount(sadj[e]);
            s_pairs /= 2;
            long all_pairs = (long)m * (m - 1) / 2;
            if (s_pairs == all_pairs) {
                /* L²(G) complete → χ'_s = m */
                passed = (m <= bound);
                path = "shortcut";
                path_shortcut++;
            } else {
                int r = multi_shot_greedy(m, sadj, bound, shots);
                if (r > 0 && r <= bound) {
                    passed = 1; path = "greedy"; path_greedy++;
                } else {
                    passed = 0; path = "unresolved"; path_unresolved++;
                }
            }
        }
        if (csv) {
            printf("%s,%d,%d,%d,%s\n", line, n, m, bound,
                   passed ? path : "FAIL_unresolved");
        }
        if (!passed) {
            candidates++;
            fprintf(stderr, "!!! CANDIDATE COUNTEREXAMPLE: g6=%s n=%d m=%d "
                    "(χ'_s > %d via path=%s) !!!\n", line, n, m, bound, path);
        }
        count++;
        if (count % every == 0) {
            clock_gettime(CLOCK_MONOTONIC, &t1);
            double dt = (t1.tv_sec - t0.tv_sec)
                      + (t1.tv_nsec - t0.tv_nsec) / 1e9;
            fprintf(stderr, "# %ld graphs done, %ld candidates, "
                    "elapsed=%.1fs, rate=%.0fg/s\n",
                    count, candidates, dt, count / (dt > 0 ? dt : 1));
        }
    }
    clock_gettime(CLOCK_MONOTONIC, &t1);
    double dt = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;
    fprintf(stderr, "# done: %ld graphs in %.1fs (rate %.0fg/s)\n",
            count, dt, count / (dt > 0 ? dt : 1));
    fprintf(stderr, "# decision paths: shortcut=%ld greedy=%ld unresolved=%ld\n",
            path_shortcut, path_greedy, path_unresolved);
    if (candidates > 0)
        fprintf(stderr, "# %ld CANDIDATE COUNTEREXAMPLES (chi_s > %d)\n",
                candidates, bound);
    else
        fprintf(stderr, "# NO counterexamples found (all chi_s ≤ %d)\n", bound);
    return candidates > 0 ? 0 : 0;
}
